snr.py

```diff
@@ -0,0 +1,287 @@
+                           ad_err=np.sqrt(0.289) * u.adu / u.pixel):
+    """
+    Returns the exposure time needed (in seconds) to achieve the desired
+    signal to noise ratio (from "Handbook of CCD Astronomy", Steve Howell,
+    2000, pg 55).
+
+    Parameters
+    ----------
+    snr : `~astropy.units.Quantity`
+        The signal to noise ratio of the given observation in untis of
+        sqrt(electrons).
+    countrate : `~astropy.units.Quantity`
+        The counts per second with units of electron/second.
+    npix : `~astropy.units.Quantity`, optional
+        Number of pixels under consideration for the signal with units of
+        pixels. Default is 1 * astropy.units.pixel.
+    n_background : `~astropy.units.Quantity`, optional
+        Number of pixels used in the background estimation with units of
+        pixels. Default is set to np.inf * astropy.units.pixel
+        such that there is no contribution of error due to background
+        estimation. This assumes that n_background will be >> npix.
+    background_rate : `~astropy.units.Quantity`, optional
+        Photons per pixel per second due to the backround/sky with units of
+        electrons/second/pixel.
+        Default is 0 * (astropy.units.electron /
+                        astropy.units.second / astropy.units.pixel)
+    darkcurrent_rate : `~astropy.units.Quantity`, optional
+        Electrons per pixel per second due to the dark current with units
+        of electrons/second/pixel.
+        Default is 0 * (astropy.units.electron /
+                        astropy.units.second / astropy.units.pixel)
+    readnoise : `~astropy.units.Quantity`, optional
+        Electrons per pixel from the read noise with units of electrons/pixel.
+        Default is 0 * astropy.units.electron / astropy.units.pixel.
+    gain : `~astropy.units.Quantity`, optional
+        Gain of the CCD with units of electrons/ADU. Default is
+        1 * astropy.units.electron / astropy.units.adu such that the
+        contribution to the error due to the gain is assumed to be small.
+    ad_err : `~astropy.units.Quantity`, optional
+        An estimate of the 1 sigma error within the A/D converter with units of
+        adu/pixel. Default is set to
+        0.289 * astropy.units.adu / astropy.units.pixel
+        (Merline & Howell, 1995).
+
+    Returns
+    -------
+    t : `~astropy.units.Quantity`
+        The exposure time needed (in seconds) to achieve the given signal
+        to noise ratio.
+    """
+    readnoise = get_shotnoise(readnoise)
+    gain_err = get_shotnoise(gain * ad_err)
+
+    # solve t with the quadratic equation (pg. 57 of Howell 2000)
+    A = countrate ** 2
+    B = (-1) * snr ** 2 * (countrate + npix * (background_rate +
+                                               darkcurrent_rate))
+    C = (-1) * snr ** 2 * npix * readnoise ** 2
+
+    t = (-B + np.sqrt(B ** 2 - 4 * A * C)) / (2 * A)
+
+    if gain_err.value > 1 or n_background.value != np.inf:
```

Comment:
Two things here: 
* wherever you strip the units off of a value, first explicitly convert the quantity into the correct units. So instead of `gain_err.value` -> `gain_err.to(u.electron / u.adu).value`. Otherwise the units could be different than what you're assuming (because the decorator is actually checking that the unit of the input argument is _convertable into_ the unit specified in the decorator, not that it is precisely _equivalent to_ the unit in the decorator)
* When comparing infinities, it's safest to use `not np.isfinite(<value>)` instead of `<value> != np.inf`

Reply:
`np.isfinite` also excludes `np.nan`, do you want `np.nan` to pass through or not?

Reply:
Personally I like that behavior, do you think it makes sense here?

Reply:
Yes, I agree with you.

Reply:
> * When comparing infinities, it's safest to use `not np.isfinite(<value>)` instead of `<value> != np.inf`

Out of curiosity, why is `not np.isfinite(<value>)` safer than `<value> != np.inf`?

Reply:
It doesn't care about the sign of the infinity, and it passes nans. I tend to prefer functions that will still run when you give them bad inputs (and return nans) rather than failing with obscure errors like `<value> == np.inf`.

Reply:
Don't forget negative infinity. Maybe this will make it clearer:

https://docs.scipy.org/doc/numpy/reference/generated/numpy.isinf.html

Reply:
And https://docs.scipy.org/doc/numpy/reference/generated/numpy.isfinite.html

Reply:
> (because the decorator is actually checking that the unit of the input argument is _convertable into_ the unit specified in the decorator, not that it is precisely _equivalent to_ the unit in the decorator)

This is good to know, I was thinking that it checks that they are equivalent.

> wherever you strip the units off of a value, first explicitly convert the quantity into the correct units. So instead of `gain_err.value` -> `gain_err.to(u.electron / u.adu).value`. Otherwise the units could be different than what you're assuming

I think since `gain_err = get_shotnoise(gain * ad_err)` where the function of `get_shotnoise()` is to force units, I'll explicitly convert the quantity in the call to `get_shotnoise()` instead of in the conditional statement. So I'll have something like
```
# get units in sqrt(u.electron / u.pixel)
gain_err = get_shotnoise(gain.to(u.electron / u.adu) * ad_err.to(u.adu / u.pixel))
# now gain_err definitely has the right units
.
.
.
if gain_err.value > 1 or not np.isfinite(n_background.value):
```
I think that does the same thing but also ensures `gain_err` truly has the correct units.

Reply:
Actually it would make  more sense to convert units inside of `get_shotnoise()`... either way it would always return `gain_err` in the correct units